npm/webpack-preprocessor/lib/cross-origin-callback-loader.ts

```diff
@@ -0,0 +1,165 @@
+      allowSuperOutsideMethod: true,
+      allowUndeclaredExports: true,
+      sourceType: 'unambiguous',
+    })
+  } catch (err) {
+    // it's unlikely there will be a parsing error, since that should have
+    // already been caught by a previous loader, but if there is and it isn't
+    // possible to get the AST, there's nothing we can do, so just return the
+    // original source
+    debug('parsing error for file (%s): %s', resourcePath, err.stack)
+
+    return source
+  }
+
+  let hasDependencies = false
+
+  traverse(ast, {
+    CallExpression (path) {
+      const callee = path.get('callee') as NodePath<t.MemberExpression>
+
+      if (!callee.isMemberExpression()) return
+
+      // check if we're inside a cy.origin() callback
+      if ((callee.node.property as t.Identifier).name !== 'origin') return
+
+      const lastArg = _.last(path.get('arguments'))
+
+      // the user could try an invalid signature for cy.origin() where the
+      // last argument is not a function. in this case, we'll return the
+      // unmodified code and it will be a runtime validation error
+      if (
+        !lastArg || (
+          !lastArg.isArrowFunctionExpression()
+          && !lastArg.isFunctionExpression()
+        )
+      ) {
+        return
+      }
+
+      // replace instances of Cypress.require('dep') with require('dep')
+      lastArg.traverse({
+        CallExpression (path) {
+          const callee = path.get('callee') as NodePath<t.MemberExpression>
+
+          // e.g. const dep = Cypress.require('../path/to/dep')
+          if (callee.matchesPattern('Cypress.require')) {
+            hasDependencies = true
+
+            path.replaceWith(
+              t.callExpression(
+                callee.node.property as t.Expression, // 'require'
+                path.get('arguments').map((arg) => arg.node), // ['../path/to/dep']
+              ),
+            )
+          }
+        },
+      }, this)
+
+      if (!hasDependencies) return
+
+      // generate the extracted callback function from an AST into a string
+      // and assign it to a variable. we wrap this generated code when we
+      // eval the code, so the variable is set up and then invoked. it ends up
+      // like this:
+      //
+      // let __cypressCrossOriginCallback              】added at runtime
+      // (function () {                                ┓ added by webpack
+      //   // ... webpack stuff stuff ...              ┛
+      //   __cypressCrossOriginCallback = (args) => {  ┓ extracted callback
+      //     const dep = require('../path/to/dep')     ┃
+      //     // ... test stuff ...                     ┃
+      //   }                                           ┛
+      //   // ... webpack stuff stuff ...              ┓ added by webpack
+      // }())                                          ┛
+      // __cypressCrossOriginCallback(args)            】added at runtime
+      //
+      const callbackName = '__cypressCrossOriginCallback'
```

Comment:
Once we start on the iframe handling, will we be reusing this?

Reply:
Yes, we will.